server/sonar-server/src/main/java/org/sonar/server/permission/GroupPermissionChanger.java

```diff
@@ -0,0 +1,94 @@
+import org.sonar.db.DbClient;
+import org.sonar.db.DbSession;
+import org.sonar.db.permission.GroupPermissionDto;
+import org.sonar.server.exceptions.BadRequestException;
+import org.sonar.server.user.UserSession;
+
+import static org.sonar.server.permission.ws.PermissionRequestValidator.validateNotAnyoneAndAdminPermission;
+
+public class GroupPermissionChanger {
+
+  private final DbClient dbClient;
+  private final UserSession userSession;
+
+  public GroupPermissionChanger(DbClient dbClient, UserSession userSession) {
+    this.dbClient = dbClient;
+    this.userSession = userSession;
+  }
+
+  public boolean apply(DbSession dbSession, GroupPermissionChange change) {
+    PermissionPrivilegeChecker.checkProjectAdminUserByComponentUuid(userSession, change.getProjectUuid());
+
+    if (shouldSkip(dbSession, change)) {
+      return false;
+    }
+
+    switch (change.getOperation()) {
+      case ADD:
+        validateNotAnyoneAndAdminPermission(change.getPermission(), change.getGroupIdOrAnyone());
+        GroupPermissionDto addedDto = new GroupPermissionDto()
+          .setRole(change.getPermission())
+          // TODO support organizations
+          .setGroupId(change.getGroupIdOrAnyone().getId())
+          .setResourceId(change.getProjectId());
+        dbClient.groupPermissionDao().insert(dbSession, addedDto);
+        break;
+      case REMOVE:
+        checkAdminUsersExistOutsideTheRemovedGroup(dbSession, change);
+        GroupPermissionDto deletedDto = new GroupPermissionDto()
+          .setRole(change.getPermission())
+          // TODO support organizations
+          .setGroupId(change.getGroupIdOrAnyone().getId())
+          .setResourceId(change.getProjectId());
+        dbClient.roleDao().deleteGroupRole(deletedDto, dbSession);
+        break;
+      default:
+        throw new UnsupportedOperationException("Unsupported permission change: " + change.getOperation());
+    }
+    return true;
+  }
+
+  private boolean shouldSkip(DbSession dbSession, GroupPermissionChange change) {
+    List<String> existingPermissions;
+    if (change.getGroupIdOrAnyone().isAnyone()) {
+      existingPermissions = dbClient.groupPermissionDao().selectAnyonePermissions(dbSession, change.getProjectId());
+    } else {
+      existingPermissions = dbClient.groupPermissionDao().selectGroupPermissions(dbSession, change.getGroupIdOrAnyone().getId(), change.getProjectId());
+    }
+    return (PermissionChange.Operation.ADD == change.getOperation() && existingPermissions.contains(change.getPermission())) ||
+      (PermissionChange.Operation.REMOVE == change.getOperation() && !existingPermissions.contains(change.getPermission()));
+  }
+
+  private void checkAdminUsersExistOutsideTheRemovedGroup(DbSession dbSession, GroupPermissionChange change) {
+    if (GlobalPermissions.SYSTEM_ADMIN.equals(change.getPermission()) &&
+      !change.getProjectRef().isPresent() &&
+      // TODO support organizations
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1135)